crates/core/src/factories/core_kitsune.rs

```diff
@@ -170,6 +170,101 @@ impl Kitsune for CoreKitsune {
         })
     }
 
+    fn remove_space(&self, space: SpaceId) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            let fut = self.map.lock().unwrap().get(&space).cloned();
+            if let Some(fut) = fut {
+                if let Ok(s) = fut.await {
+                    // Just a point in time check, try to help the user out if they accidentally
+                    // try to remove a space with active local agents.
+                    // Those agents need to leave the space before it should be removed.
+                    if !s.local_agent_store().get_all().await?.is_empty() {
+                        return Err(K2Error::other(
+                            "Cannot remove space with local agents",
+                        ));
+                    }
+
+                    // Checks passed, remove our reference to the space.
+                    self.map.lock().unwrap().remove(&space);
+
+                    // Unregister the space and its module handlers from the transport.
+                    self.transport().await?.unregister_space(space).await;
+
+                    // Get all the peer URLs of connected peers.
+                    let connected_peer_urls = Arc::new(RwLock::new(
+                        self.transport()
+                            .await?
+                            .get_connected_peers()
+                            .await?
+                            .into_iter()
+                            .collect::<HashSet<_>>(),
+                    ));
+
+                    // Remove any peer URLs that are connected for another space.
+                    let spaces = self
+                        .map
+                        .lock()
+                        .unwrap()
+                        .iter()
+                        .map(|s| s.1.clone())
+                        .collect::<Vec<_>>();
+                    futures::future::join_all(spaces.into_iter().map(
+                        |fut| -> BoxFut<'_, K2Result<()>> {
+                            let connected_peer_urls =
+                                connected_peer_urls.clone();
+                            Box::pin(async move {
```

Comment:
Can this just be an `async` block instead of a boxed future?

Reply:
I don't *think* so. The return type `impl Future` isn't allowed by Rust and I can't make an `async` closure. I could define an actual async function to do this work and that might work but even then, I think Rust complains about the futures having different types when you try to put them in a collection and you end up needing to Box anyway